src/sentry/taskworker/pending_task_store.py

```diff
@@ -272,6 +277,433 @@ def remove_completed(self) -> None:
+        task_ids = self.client.zrevrange(status_key, 0, 0)
+        if not task_ids:
+            return None
+        return task_ids[0]
+
+    def _get_all_tasks_with_status(self, status: int) -> list[str]:
+        status_key = self._build_inflight_activations_key(status)
+        tasks = self.client.zrange(status_key, 0, -1)
+        return tasks
+
+    def _get_score(self, status: int, task_id: str) -> int | None:
+        status_key = self._build_inflight_activations_key(status)
+        return int(self.client.zscore(status_key, task_id))
+
+    def _count_all_offset_activations(self, status: int) -> int:
+        key = self._build_inflight_activations_key(status)
+        return self.client.zcard(key)
+
+    def _remove_inflight_activation(self, task_id: str) -> None:
+        # TODO: is there a better way to do this?
+        fields = list(self._get_all_inflight_activation_fields(task_id).keys())
+        if not fields:
+            return
+        self._remove_inflight_activation_field(task_id, *fields)
+
+    def _remove_inflight_activation_field(self, task_id: str, *fields) -> None:
+        key = self._build_inflight_activation_key(task_id)
+        self.client.hdel(key, *fields)
+
+    def _remove_inflight_activations(self, status: int, task_id: str) -> None:
+        key = self._build_inflight_activations_key(status)
+        self.client.zrem(key, task_id)
+
+    def _remove_task(self, task_id: str) -> None:
+        key = self._build_task_key(task_id)
+        self.client.delete(key)
+
+    def _compose_inflight_activation_proto(self, task_id: str) -> InflightActivation:
+        task_json = self._get_task(task_id)
+        task_proto = Parse(task_json, TaskActivation())
+        inflight_dict = self._get_all_inflight_activation_fields(task_id)
+        inflight_proto = ParseDict(inflight_dict, InflightActivation())
+        inflight_proto.activation.CopyFrom(task_proto)
+        return inflight_proto
+
+    def store(self, batch: Sequence[InflightActivation]):
+        for inflight in batch:
+            task_id = inflight.activation.id
+            status = inflight.status  # status is an integer
+            offset = inflight.offset
+            inflight_dict = MessageToDict(
+                inflight, use_integers_for_enums=True, preserving_proto_field_name=True
+            )
+            task_json = MessageToJson(inflight.activation)
+
+            self._set_task(task_id, task_json)
+            del inflight_dict[
+                "activation"
+            ]  # remove the activation field as it is stored in a separate key
+            self._set_inflight_activation_attributes(task_id, inflight_dict)
+            self._set_inflight_activations(status, offset, task_id)
+
+    def get_pending_task(self) -> InflightActivation | None:
+        # TODO: can maybe use zpopmin instead here?
+        task_id = self._get_min_offset_activation(TASK_ACTIVATION_STATUS_PENDING)
```

Comment:
zpopmin could help races and duplicate task execution if that shows up in the test harness.